pl.from_numpy produces column with null dtype when input array is empty

### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import numpy as np
import polars as pl
from polars.testing import assert_frame_equal

data = np.zeros((0,))
assert_frame_equal(pl.from_numpy(data, schema=["foo"]), pl.Series("foo", data).to_frame())
```


### Log output

```shell
Traceback (most recent call last):
  File "repro.py", line 6, in <module>
    assert_frame_equal(pl.from_numpy(data, schema=["foo"]), pl.Series("foo", data).to_frame())
  File ".venv/lib/python3.11/site-packages/polars/_utils/deprecation.py", line 91, in wrapper
    return function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".venv/lib/python3.11/site-packages/polars/testing/asserts/frame.py", line 91, in assert_frame_equal
    _assert_frame_schema_equal(
  File ".venv/lib/python3.11/site-packages/polars/testing/asserts/frame.py", line 187, in _assert_frame_schema_equal
    raise_assertion_error(objects, detail, left_schema_dict, right_schema_dict)
  File ".venv/lib/python3.11/site-packages/polars/testing/asserts/utils.py", line 12, in raise_assertion_error
    raise AssertionError(msg) from cause
AssertionError: DataFrames are different (dtypes do not match)
[left]:  {'foo': Null}
[right]: {'foo': Float64}
```


### Issue description

If you change the shape of `data` to `(1,)` then the assertion passes

### Expected behavior

Assertion passes when the shape of the data is `(0,)`

### Installed versions

<details>

```
Polars:               1.5.0
Index type:           UInt32
Platform:             Linux-6.5.0-44-generic-x86_64-with-glibc2.35
Python:               3.11.9 (main, Apr  6 2024, 17:59:24) [GCC 11.4.0]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           <not installed>
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
great_tables:         <not installed>
hvplot:               <not installed>
matplotlib:           <not installed>
nest_asyncio:         <not installed>
numpy:                2.1.0
openpyxl:             <not installed>
pandas:               <not installed>
pyarrow:              <not installed>
pydantic:             <not installed>
pyiceberg:            <not installed>
sqlalchemy:           <not installed>
torch:                <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

--- a/py-polars/polars/_utils/construction/dataframe.py
+++ b/py-polars/polars/_utils/construction/dataframe.py
@@ -1289,7 +1289,7 @@ def numpy_to_pydf(
             )._s
             for series_name, record_name in zip(column_names, record_names)
         ]
-    elif shape == (0,):
+    elif shape == (0,) and n_columns == 0:
         data_series = []
 
     elif len(shape) == 1:

--- a/py-polars/tests/unit/interop/numpy/test_from_numpy_df.py
+++ b/py-polars/tests/unit/interop/numpy/test_from_numpy_df.py
@@ -10,7 +10,9 @@ import polars as pl
 from polars.testing import assert_frame_equal
 
 if TYPE_CHECKING:
-    from polars._typing import PolarsTemporalType
+    import numpy.typing as npt
+
+    from polars._typing import PolarsDataType, PolarsTemporalType
 
 
 def test_from_numpy() -> None:
@@ -156,7 +158,7 @@ def test_from_numpy_supported_units(
         (np.int32, pl.Int32),
     ],
 )
-def test_from_numpy_empty(np_dtype: np.dtype, dtype: pl.DataType) -> None:
+def test_from_numpy_empty(np_dtype: npt.DTypeLike, dtype: PolarsDataType) -> None:
     data = np.array([], dtype=np_dtype)
     result = pl.from_numpy(data, schema=["a"])
     expected = pl.Series("a", [], dtype=dtype).to_frame()

--- a/py-polars/tests/unit/interop/numpy/test_from_numpy_df.py
+++ b/py-polars/tests/unit/interop/numpy/test_from_numpy_df.py
@@ -147,3 +147,17 @@ def test_from_numpy_supported_units(
         pl.Series("column_0", expected_values).str.strptime(expected_dtype).to_frame()
     )
     assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize(
+    ("np_dtype", "dtype"),
+    [
+        (np.float64, pl.Float64),
+        (np.int32, pl.Int32),
+    ],
+)
+def test_from_numpy_empty(np_dtype: np.dtype, dtype: pl.DataType) -> None:
+    data = np.array([], dtype=np_dtype)
+    result = pl.from_numpy(data, schema=["a"])
+    expected = pl.Series("a", [], dtype=dtype).to_frame()
+    assert_frame_equal(result, expected)

